### DISCUSSION TITLE
Micronaut Netty client handler "micronaut-http-response" not available when customizing channel pipeline for logbook handler

### DISCUSSION THREAD

**Discussion by subash89** [micronaut-projects/micronaut-core#10524] [Q&A]
I am trying to add logbook client handler using NettyClientCustomizer.Registry bean listener as per micronaut docs.
But getting following error.  Reason is by the time I am suppose to register my handler, ChannelPipelineCustomizer.HANDLER_MICRONAUT_HTTP_RESPONSE not available. Any suggestions to fix this ?

```
  @Override
        public void onStreamPipelineBuilt() {
            channel.pipeline().addBefore(
                    ChannelPipelineCustomizer.HANDLER_MICRONAUT_HTTP_RESPONSE,
                    "logbook-client",
                    new LogbookClientHandler(logbook)
            );
        }
```

```
java.util.NoSuchElementException: micronaut-http-response
	at io.netty.channel.DefaultChannelPipeline.getContextOrDie(DefaultChannelPipeline.java:1073)
	at io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:248)
	at io.netty.channel.DefaultChannelPipeline.addBefore(DefaultChannelPipeline.java:237)
	at com.test.logging.logbook.LogbookClientCustomizer$Customizer.onStreamPipelineBuilt(LogbookClientCustomizer.java:75)
	at io.micronaut.http.netty.AbstractCompositeCustomizer.forEach(AbstractCompositeCustomizer.java:109)
	at io.micronaut.http.client.netty.CompositeNettyClientCustomizer.onStreamPipelineBuilt(CompositeNettyClientCustomizer.java:57)
	at io.micronaut.http.client.netty.ConnectionManager$Pool$Http2ConnectionHolder.init(ConnectionManager.java:1366)
	at io.micronaut.http.client.netty.ConnectionManager$4.channelRead(ConnectionManager.java:622)
```

**Comment by yawkat** [comment]
same as #10514

**Reply by subash89** [reply]
@yawkat  any workaround you know of to place our handler in this position ?

**Reply by yawkat** [reply]
look at the pipeline with a debugger and see where the handler fits in. do this for http, https, and https+http/2 too.

**Reply by subash89** [reply]
@yawkat  By the time NettyClientCustomizer.Registry  bean listener get invoked, client pipeline head is "http2-connection".  I can do (I tried) add last or addAfter("http2-connection"). But it's no use, as logbook handler needs a full netty HttpRequest or HttpResponse to work with. Meaning this handler needs to be added some where after HttpObjectAggregator. 

But as I mentioned earlier, by the time my NettyClientCustomizer.Registry  bean listener invoked to register the channel handler, HttpObjectAggregator or "micronaut-http-response" handler not registered. 

**Is there any way I can register a channel handler after "micronaut-http-response" ?**

**Reply by yawkat** [reply]
NettyClientCustomizer should be invoked multiple times. sounds like you should override onRequestPipelineBuilt.

Note that the logbook integration will stop working with http 2 soon.

**Reply by subash89** [reply]
@yawkat we need http2 support. So to log message body, what other recommended alternatives we have in micronaut ?

**Reply by yawkat** [reply]
you can use a filter with `@Body`, but it will mean buffering the body fully all the time

**Reply by subash89** [reply]
The main reason is the loggings and obfuscations logbook provides. Can you explain a bit more why logbook support stop working with http2  ? And any alternative for logbook functionality will be provided ?

**Reply by yawkat** [reply]
as of (probably) 4.5.0, the http2 stack will not use per-stream pipelines anymore, and will not use intermediate http1 objects on the pipeline. This is for performance reasons. Both of these things were necessary for the logbook netty integration.

It will be necessary to build a logbook integration based on filters. Maybe it would be worthwhile to build a dedicated micronaut-logbook module for this use case.

**Reply by subash89** [reply]
I tried using @Body in the server filter. I could read (@Body String ) in @RequestFilter. But in Server @ResponseFilter, I cannot inject @Body. only option was to do something like   response.getBody() in my method.  But in here this getting response filter engaged before response getting serialized. response.getBody actually returns my json model  PoJo class. What should I do to read response body as a String.

**Reply by subash89** [reply]
@yawkat  the question is more like on Response path, is there a way I can access to the serialized(String) version of the response body.  I tried lowest precedence for my response server filter method, but still when I do response.getBody() it still has the Json PoJo object.
